Feat / Geo blocking for media items

langemike opened the pull request:
This change will show a nice message to the user when a media item is geographically blocked. Previously you would see a generic JW Player error message on top of the player.

**EDIT: Updated screenshots on 11 january**

### Cinema example
![Screenshot 2024-01-11 at 14 21 37](https://github.com/Videodock/ott-web-app/assets/1019129/4937682c-b2af-433c-82ab-bde57ecc389b)

### Inline player example
![Screenshot 2024-01-11 at 14 22 00](https://github.com/Videodock/ott-web-app/assets/1019129/7da5ab0c-28dc-4735-875b-c2fd475d5f46)

It will determine the availability of the media item based on the HTTP statuscode of the m3u8 (playlist) file based on a `HEAD`-request. The media API itself will always return a 200 statuscode. 

I created a new `<PlayerError />` component which handles the display of the geo blocked message. In the future this could be extended to also show other (nice) error messages instead of the (default) JWPlayer error box.

I noticed that for a split second it can occur that you will see a JWPlayer error box which will get replaced by this nice friendly message. If this is something we do not want to occur, I will need to invest some extra time in it. ❗ 

I also changed the cinema title to a `<h1>` to keep the heading hierarchy logical.

I have also tested this on a iOS device. Responsiveness is also guaranteed.

Tickets: 
- https://videodock.atlassian.net/browse/OTT-363 - Generic logic
- https://videodock.atlassian.net/browse/OTT-364 - Implementation
- https://videodock.atlassian.net/browse/OTT-365 - UI & translation
- https://videodock.atlassian.net/browse/OTT-366 - Testing

langemike commented:
I noticed a couple of tests (unit, e2e) failed. I will fix those a.s.a.p.

royschut commented on platforms/web/public/locales/en/video.json:
```diff
@@ -12,6 +12,11 @@
   "live": "Live",
+  "player_error": {
+    "geo_blocked_description": "We apologize, but this content is currently unavailable in your region.",
+    "geo_blocked_title": "Content Unavailable in your region",
```
Perhaps a shorter title looks better on mobile? Also no cap `u`. :-) 
```suggestion
    "geo_blocked_title": "Content unavailable",
```

langemike replied:
Changed!

royschut reviewed (commented):
Although I like the approach you've chosen, I'm not sure about moving the logics one level up. Ideally the PlayerContainer remains responsible for all logics. I have an alternative approach in mind. let me try to explain:
- Move the hook implentations back to PlayerContainer
- Make a PlayerErrorComponent with all the jsx and styling that you've created, for now focused on the 'geo' state. 
- Maybe we can try to make both player error views the same, to keep things more simple. We could choose to always, or never, show the logo. The 'go back' button might not be necessary if we just keep the Cinema back button.
- To make sure that the other overlay elements don't show in Cinema, perhaps we can add an onPlayerError callback prop to PlayerContainer, with a handler in Cinema that simply keeps a state of the error, and (partially) hides the overlay.

langemike commented:
> Although I like the approach you've chosen, I'm not sure about moving the logics one level up. Ideally the PlayerContainer remains responsible for all logics. I have an alternative approach in mind. let me try to explain:
> 
> * Move the hook implentations back to PlayerContainer
> * Make a PlayerErrorComponent with all the jsx and styling that you've created, for now focused on the 'geo' state.
> * Maybe we can try to make both player error views the same, to keep things more simple. We could choose to always, or never, show the logo. The 'go back' button might not be necessary if we just keep the Cinema back button.
> * To make sure that the other overlay elements don't show in Cinema, perhaps we can add an onPlayerError callback prop to PlayerContainer, with a handler in Cinema that simply keeps a state of the error, and (partially) hides the overlay.

Thanks for your insightfull feedback! Sounds like a good approach I haven't thought of.

ChristiaanScheermeijer reviewed (commented):
I agree with the feedback given by Roy (we have discussed this yesterday together)

langemike committed:
feat(player): show nice message when accessing a geo blocked media item

langemike commented:
@ChristiaanScheermeijer & @royschut I have implemented the feedback, force pushed my changes to keep the git history clean and rewrote the PR description based on the new situation. I'm hungry for feedback

royschut commented on packages/hooks-react/src/useProtectedMedia.ts:
```diff
@@ -7,5 +8,20 @@ import useContentProtection from './useContentProtection';
 
-  return useContentProtection('media', item.mediaid, (token, drmPolicyId) => apiService.getMediaById(item.mediaid, token, drmPolicyId));
+  const [isGeoBlocked, setIsGeoBlocked] = useState(false);
+  const queryResult = useContentProtection('media', item.mediaid, (token, drmPolicyId) => apiService.getMediaById(item.mediaid, token, drmPolicyId));
```
Minor suggestion (`queryResult` is a bit generic):
```suggestion
  const contentProtectionQuery = useContentProtection('media', item.mediaid, (token, drmPolicyId) => apiService.getMediaById(item.mediaid, token, drmPolicyId));
```

royschut commented on packages/hooks-react/src/useProtectedMedia.ts:
```diff
@@ -7,5 +8,20 @@ import useContentProtection from './useContentProtection';
+  const queryResult = useContentProtection('media', item.mediaid, (token, drmPolicyId) => apiService.getMediaById(item.mediaid, token, drmPolicyId));
+
+  useEffect(() => {
+    const m3u8 = queryResult.data?.sources.find((source) => source.file.indexOf('.m3u8') !== -1);
```
Maybe add an early return on `!contentProtectionQuery.isSuccess`

langemike replied:
I think this is not needed, because queryResult.data will not change it's value and therefore the effect will not be called. Also added that the optional chaining will cause the `fetch` to be skipped ;)

royschut reviewed (approved):
Like it! Very clean.
I see you've kept the Cinema overlay, but I don't think it's really a problem indeed.

If you wanna fix temporary seeing the player error, then you could make isGeoBlocked first true, and set it to false after the call has finished or not necessary. Or better even, reverse it: `const [hasGeoAccess, setHasGeoAccess] = useState(false);` (combined with another loading state). But to be honest, I don't think it's important enough. :-)

langemike committed:
chore: change variable name

langemike commented:
> If you wanna fix temporary seeing the player error

Thanks for the suggestions. I'm a bit hesitant about the approaches. Maybe changing the `isLoaded` state based on the HEAD-request if a m3u8 is found? That way the code in PlayerContainer can be the same.

Since it's something it rarely occurs, I think better leave it as-is.

royschut commented:
> I think better leave it as-is

Yes, agreed!

ChristiaanScheermeijer reviewed (approved):
Nice @langemike! We had some ideas to make the geo check a bit more abstract, but let's do that with the refactoring of the services & hooks.
